flocker/acceptance/testtools.py

```diff
@@ -869,24 +872,56 @@ def failed_query(failure):
                        failed_query)
         return d
     agents_connected = loop_until(reactor, nodes_available)
+    agents_connected.addCallback(lambda _: _add_nodes(cluster))
+    return agents_connected
+
+
+def _add_nodes(cluster):
+    """
+    Configure the ``Node`` objects for a newly created ``Cluster`` whose
+    nodes are known to be alive.
+
+    :param Cluster cluster: Cluster that still needs nodes set.
+    :return: ``cluster`` updated with appropriate ``nodes`` set.
+    """
+    # By default we just trust address returned by Flocker
+    def default_get_public_ip(address):
+        return address
 
-    # Extract node hostnames from API that lists nodes. Currently we
-    # happen know these in advance, but in FLOC-1631 node identification
-    # will switch to UUIDs instead.
-    agents_connected.addCallback(lambda _: cluster.current_nodes())
+    try:
+        backend = get_backend_api(None, cluster.cluster_uuid)
+    except SkipTest:
+        # Can't load backend, will have to trust Flocker's reported IPs.
+        get_public_ip = default_get_public_ip
+    else:
+        if ICloudAPI.providedBy(backend):
+            node_ips = list(set(ip_address(i) for i in ips)
+                            for ips in backend.list_live_nodes().values())
+
+            def get_public_ip(address):
+                for ips in node_ips:
+                    if ip_address(address) in ips:
+                        return [unicode(ip) for ip in ips
+                                if not any([ip.is_private, ip.is_link_local,
+                                            ip.is_loopback])][0]
+                raise ValueError(
+                    "Couldn't find address in cloud API reported IPs")
+        else:
+            get_public_ip = default_get_public_ip
 
     def node_from_dict(node):
         reported_hostname = node["host"]
-        public_address = hostname_to_public_address.get(
-            reported_hostname, reported_hostname)
+        public_address = get_public_ip(reported_hostname)
```

Comment:
:grey_question: This is kinda gross. This assumes that `reported_hostname` is a unicode IP address, and will blow up if it is not. Which is bonus fun because it's named "hostname" and not "ip_address".

Perhaps okay since this is test code, and it blows up... but still.